### PR TITLE
Add Flowers102 dataset

### PR DESCRIPTION
**Pull request by zhiqwang**
Addresses Flowers102 dataset in #5108.

cc @pmeier @NicolasHug

### CODE REVIEW

File: torchvision/datasets/flowers102.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        transform (callable, optional): A function/transform that takes in an PIL image and returns a
+            transformed version. E.g, ``transforms.RandomCrop``.
+        target_transform (callable, optional): A function/transform that takes in the target and transforms it.
+    """
+
+    def __init__(
+        self,
+        root: str,
+        split: str = "train",
+        download: bool = True,
```
Comment (pmeier) [inline torchvision/datasets/flowers102.py:37]: The download flag is missing from the docstring.

File: torchvision/datasets/flowers102.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        transform: Optional[Callable] = None,
+        target_transform: Optional[Callable] = None,
+    ) -> None:
+        super().__init__(root, transform=transform, target_transform=target_transform)
+        self._split = verify_str_arg(split, "split", ("train", "valid", "test"))
+        self._base_folder = Path(self.root) / "flowers-102"
+        self._meta_folder = self._base_folder / "labels"
```
Comment (pmeier) [inline torchvision/datasets/flowers102.py:44]: Do we actually need this? The images come in their separate folder anyway, so I feel like we can place the two other files in `self.root`.

File: torchvision/datasets/flowers102.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        self._images_folder = self._base_folder / "jpg"
+
+        if download:
+            self._download()
+
+        if not self._check_exists():
+            raise RuntimeError("Dataset not found. You can use download=True to download it")
+
+        self._labels = []
+        self._image_files = []
```
Comment (pmeier) [inline torchvision/datasets/flowers102.py:54]: There is no need to create them up here. Please move them down below to where they are actually used.

File: torchvision/datasets/flowers102.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+        from scipy.io import loadmat
+
+        # Read the label ids
+        label_mat = loadmat(self._meta_folder / "imagelabels.mat")
+        labels = label_mat["labels"][0]
+
+        self.classes = np.unique(labels).tolist()
+        self.class_to_idx = dict(zip(self.classes, range(len(self.classes))))
```
Comment (pmeier) [inline torchvision/datasets/flowers102.py:63]: `self.classes` is traditionally reserved for human-readable names. What you define here would be more fitting for `self.labels`.

If I'm not mistaken, there is no way to get the human readable classes from the original data. So we could either use some external mapping such as [this](https://gist.github.com/JosephKJ/94c7728ed1a8e0cd87fe6a029769cde1) or [this](https://github.com/bdevnani3/oxfordflowers102-label-name-mapping/blob/main/mapping.json), but then we would also need to make sure that this is correct and from a quick glance, both variants are not the same. The other option is to simply drop the human-readable classes outright, since the authors didn't provide them either. I would go for the latter.

File: test/test_datasets.py [hunk 5]
```diff
@@ -2463,5 +2463,45 @@ def _meta_to_split_and_classification_ann(self, meta, idx):
...
+    FEATURE_TYPES = (PIL.Image.Image, int)
+
+    ADDITIONAL_CONFIGS = datasets_utils.combinations_grid(split=("train", "valid", "test"))
+    REQUIRED_PACKAGES = ("scipy",)
+
+    def inject_fake_data(self, tmpdir: str, config):
+        base_folder = pathlib.Path(tmpdir) / "flowers-102"
+
+        num_classes = 3
+        num_images_per_split = dict(train=3, valid=3, test=4)
```
Comment (pmeier) [inline test/test_datasets.py:2477]: Let's make the number of samples different for all splits

```suggestion
        num_images_per_split = dict(train=5, valid=4, test=3)
```

**Review by pmeier — COMMENTED** [review]
Hey @zhiqwang, thanks a lot for the PR. I have some minor comments inline.

File: torchvision/datasets/flowers102.py (hunk 2 above)
Comment (zhiqwang) [reply torchvision/datasets/flowers102.py:44]: Hi @pmeier , The reason for creating this directory is to make it easier to determine if the two files were downloaded correctly.

File: torchvision/datasets/flowers102.py (hunk 1 above)
Comment (zhiqwang) [reply torchvision/datasets/flowers102.py:40]: Added now.

File: torchvision/datasets/flowers102.py (hunk 2 above)
Comment (pmeier) [reply torchvision/datasets/flowers102.py:47]: Given that we are talking about two files vs. one folder, I think we can simply check for the existence of both files? [`torchvision.datasets.utils.check_integrity`](https://github.com/pytorch/vision/blob/8265469b2526e8d520e2e5ede71454ac01e357aa/torchvision/datasets/utils.py#L66) was made for that.

File: torchvision/datasets/flowers102.py (hunk 4 above)
Comment (zhiqwang) [reply torchvision/datasets/flowers102.py]: Hi @pmeier , I rename `self.classes` to `self.labels` here.

File: torchvision/datasets/flowers102.py (hunk 2 above)
Comment (zhiqwang) [reply torchvision/datasets/flowers102.py]: Hi @pmeier , I've updated these with `check_integrity`.

File: torchvision/datasets/flowers102.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    def extra_repr(self) -> str:
+        return f"split={self._split}"
+
+    def _check_integrity(self):
+        st1 = check_integrity(str(self._base_folder / self.file_dict["image"][0]), self.file_dict["image"][1])
+        st2 = check_integrity(str(self._base_folder / self.file_dict["label"][0]), self.file_dict["label"][1])
+        st3 = check_integrity(str(self._base_folder / self.file_dict["setid"][0]), self.file_dict["setid"][1])
+        if not (st1 and st2 and st3):
+            return False
+        return True
```
Comment (pmeier) [inline torchvision/datasets/flowers102.py:107]: 1. We should not check the image archive, since the user can delete it after we have extracted it and the dataset should still work.
2. Using a loop makes this more concise without reducing readability:

```suggestion
        for id in ["label", "setid"]:
            filename, md5 = self.file_dict[id]
            if not check_integrity(str(self._base_folder / filename), md5):
                return False
        return True
```

File: torchvision/datasets/flowers102.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        if self._check_integrity():
+            print("Files already downloaded and verified")
+            return
+        download_and_extract_archive(
+            f"{self.download_url_prefix}{self.file_dict['image'][0]}",
+            str(self._base_folder),
+            md5=self.file_dict["image"][1],
+        )
+        download_url(f"{self.download_url_prefix}{self.file_dict['label'][0]}", str(self._base_folder))
+        download_url(f"{self.download_url_prefix}{self.file_dict['setid'][0]}", str(self._base_folder))
```
Comment (pmeier) [inline torchvision/datasets/flowers102.py:119]: Same as above

```suggestion
        for id in ["label", "setid"]:
            filename, md5 = self.file_dict[id]
            download_url(self.download_url_prefix + filename, str(self._base_folder), md5=md5)
```

File: torchvision/datasets/flowers102.py (hunk 4 above)
Comment (pmeier) [reply torchvision/datasets/flowers102.py]: I think you misunderstood me. As is, `self.labels` is a list of all available labels. That is not useful. Either you provide human readable strings for the classes or simply remove `self.labels` and `self.label_to_index`. Since there is no official mapping between labels and class names, I would go for the latter and remove the attributes.

**Review by pmeier — APPROVED** [review]
Thanks a lot @zhiqwang! I've pushed a minor simplification to not prolong this. LGTM from my side!

File: torchvision/datasets/flowers102.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+    def _check_integrity(self):
+        for id in ["label", "setid"]:
+            filename, md5 = self.file_dict[id]
+            if not check_integrity(str(self._base_folder / filename), md5):
+                return False
+        return True
+
+    def download(self):
+        if not self._check_exists():
```
Comment (zhiqwang) [inline torchvision/datasets/flowers102.py:105]: Hi @pmeier , I move the folder existence check here. Otherwise if the `labels` or `setids` is missing, we will download the image archive.

Comment (NicolasHug) [reply torchvision/datasets/flowers102.py:105]: Could we perhaps simplify the logic here, and just download all the files if `_check_exists()` returns False?

I'm not sure that allowing the image archive to be there without the labels or setids is really useful?

**Review by NicolasHug — COMMENTED** [review]
Thanks a lot @zhiqwang for the PR, and @pmeier for the review.

I made some very minor changes above, the only relevant one is that I changed `split="valid"` into `split="val"` for more consistency with other datasets.

This LGTM but I just have one question below regarding the download logic, LMK what you think

File: torchvision/datasets/flowers102.py (hunk 8 above)
Comment (pmeier) [reply torchvision/datasets/flowers102.py:105]: IIRC the discrepancy in `download` was introduced after my review. I agree with Nicolas here. We should have one method that checks the existence / integrity of all files and one download method that downloads everything if the former method returns `False.`

Comment (zhiqwang) [reply torchvision/datasets/flowers102.py]: Hi @pmeier and @NicolasHug , I've reverted the change back.

File: torchvision/datasets/flowers102.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        if self.target_transform:
+            label = self.target_transform(label)
+
+        return image, label
+
+    def extra_repr(self) -> str:
+        return f"split={self._split}"
+
+    def _check_integrity(self):
+        for id in ["label", "setid"]:
```
Comment (pmeier) [inline torchvision/datasets/flowers102.py:95]: We still need to check if the image folder exists

```suggestion
        if not (self._images_folder.exists() and self._images_folder.isdir()):
            return False

        for id in ["label", "setid"]:
```

**Review by NicolasHug — APPROVED** [review]
Thanks a lot @zhiqwang !! Will merge when green
